nxp_s32k/libuavcan/driver/src/s32k_libuavcan.cpp

```diff
@@ -0,0 +1,777 @@
+
+                    /* Block until module is ready */
+                    if (isSuccess(Status))
+                    {
+                        Status = S32K::flagPollTimeout_Clear(S32K::FlexCAN[i]->MCR, CAN_MCR_NOTRDY_MASK);
+                    }
+                }
+            }
+        }
+    }
+
+    /* Return status code */
+    return Status;
+}
+
+libuavcan::Result S32K_InterfaceGroup::select(libuavcan::duration::Monotonic timeout, bool ignore_write_available)
+{
+    /* Obtain timeout from object */
+    std::uint32_t cycles_timeout = static_cast<std::uint32_t>(timeout.toMicrosecond());
+
+    /* Initialization of delta variable for comparison */
+    volatile std::uint32_t delta = 0;
+
+    /* Disable LPIT channel 3 for loading */
+    LPIT0->CLRTEN |= LPIT_CLRTEN_CLR_T_EN_3(1);
+
+    /* Load LPIT with its maximum value */
+    LPIT0->TMR[3].TVAL = LPIT_TMR_CVAL_TMR_CUR_VAL_MASK;
+
+    /* Enable LPIT channel 3 for timeout start */
+    LPIT0->SETTEN |= LPIT_SETTEN_SET_T_EN_3(1);
+
+    /* Start of timed block */
+    while (delta < cycles_timeout)
+    {
+        /* Poll in each of the available interfaces */
+        for (std::uint8_t i = 0; i < S32K::CANFD_Count; i++)
+        {
+            /* Poll for available frames in RX FIFO */
+            if (!S32K::g_frame_ISRbuffer[i].empty())
+            {
+                return libuavcan::Result::Success;
+            }
+
+            /* Check for available message buffers for transmission if ignore_write_available is false */
+            else if (!ignore_write_available)
+            {
+                /* Poll the Inactive Message Buffer and Valid Priority Status flags for TX availability */
+                if ((S32K::FlexCAN[i]->ESR2 & CAN_ESR2_IMB_MASK) && (S32K::FlexCAN[i]->ESR2 & CAN_ESR2_VPS_MASK))
+                {
+                    return libuavcan::Result::Success;
+                }
+            }
+        }
+
+        /* Get current value of delta */
+        delta = LPIT_TMR_CVAL_TMR_CUR_VAL_MASK - (LPIT0->TMR[3].CVAL);
+    }
+
+    /* If this section is reached, means timeout occurred and return timeout status */
+    return libuavcan::Result::SuccessTimeout;
+}
+
+libuavcan::Result S32K_InterfaceManager::startInterfaceGroup(
+    const typename InterfaceGroupType::FrameType::Filter* filter_config,
+    std::size_t                                           filter_config_length,
+    InterfaceGroupPtrType&                                out_group)
+{
+    /* Initialize return values */
+    libuavcan::Result Status = libuavcan::Result::Success;
+    out_group                = nullptr;
+
+    /* Input validation */
+    if (filter_config_length > S32K::Filter_Count)
+    {
+        Status = libuavcan::Result::BadArgument;
+    }
+
+    /* SysClock initialization for feeding 80Mhz to FlexCAN */
+
+    /* System Oscillator (SOSC) initialization for 8Mhz external crystal */
+    SCG->SOSCCSR &= ~SCG_SOSCCSR_LK_MASK;     /* Ensure the register is unlocked */
```

Comment:
Any reasonable use of this driver will not tolerate it setting the system oscillator. You can document and respond to SCG settings and provide these settings in your examples but you cannot set them here or anywhere else in a media layer implementation.

Reply:
From page 572 of the reference manual:
![FlexCAN_clocks](https://user-images.githubusercontent.com/18625317/77843402-3ea1ae00-715a-11ea-8f56-0fad5203fb02.jpg)

The peripheral has only 2 clock sources available, SYS_CLK is the current configured one, which is synchronous, it is also the ony one that allows for descent bitrates, the other source SOSCDIV2_CLK, which is in fact asynchronous (comes from the external crystal currently set to 8Mhz) would reduce the current bitrates of 4Mbit/s in data and 1Mbit/s in nominal in at least tenfold

Reply:
> comes from the external crystal currently set to 8Mhz

This, for example, is board-dependent. Not all boards have the exact same external parts and may have other peripherals that require the clock tree to be setup differently (e.g. using an RC oscillator instead of an XOSC or using an XOSC ). This is one of the reasons that any non-trivial application uses an OS; peripherals have to share system resources and that requires compromises. For a bare-metal driver we can document the requirements and effects various clock rates will cause but you can't have one driver own the master clock on a system.

Try this to help you understand: use the S32 design studio to create a new project from the CAN-FD example. You'll see the code generated includes files that initialize the clocks and separate files that initialize the driver.

libuavcan only wants to provide the driver part. We document and check clock requirements but we don't set them.

Reply:
I agree that a driver shouldn't touch  an important setting such as the clock speed that the CPU is running from.
The mcu also has a couple of internal clock sources that are independent from the CPU, and the external oscillator:

From the reference manual page 558
![imagen](https://user-images.githubusercontent.com/18625317/77868031-ebca0400-71f6-11ea-9377-dece940edf56.png)

FIRCDIVx_CLK and SIRCDIVx_CLK are meant to feed peripherals that their setup must be independent from a selected core clock configuration, as Timers or UART for example.

But these are not available to the FlexCAN as mentioned before, the only alternative is to feed from SOCDIV2_CLK and document that  a 8Mhz crystal is assumed, which is the one that both the EVB's and the UCANS32K146 reference design boards have, with the lower bitrate tradeoff.

Reply:
We can document this requirement but should check it as well in the manager returning an error result if the clock is not set as we require. We'll want to improve on this in the future providing an ability to set different clock speeds for the bus and handling different master clock speeds. We'll capture that work as an issue in this repo though.

Reply:
I could setup for example 4 different predefined and validated bitrate speed profiles, and set a tunable constant such as the already used FIFO depth that chooses from these.

Reply:
@noxuz I've ported the S32 design studio bit timing calculation to the NuttX FlexCAN driver, which calculates the bit timing for any bitrate and clock configuration, the only thing is that you've to optimize is the sample point.

https://github.com/PetervdPerk-NXP/incubator-nuttx/blob/0713373616738deadf5e28fbb5ba9713a9b2b4e2/arch/arm/src/s32k1xx/s32k1xx_flexcan.c#L393-L512